pkg/cmd/provisioning/identity_provider.go

```diff
@@ -0,0 +1,335 @@
+func keyIDFromPublicKey(publicKey interface{}) (string, error) {
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize public key to DER format: %v", err)
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	return keyID, nil
+}
+
+// saveToFile saves the given data to a given file
+func saveToFile(fileName, filePath string, data []byte) error {
+	log.Printf("Saving %s locally at %s", fileName, filePath)
+	f, err := os.Create(filePath)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create file %s to store %s", filePath, fileName)
+	}
+	_, err = f.Write(data)
+	f.Close()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to write %s to file %s", fileName, filePath)
+	}
+	return nil
+}
+
+// isExistingIdentifyProvider checks if given identity provider is owned by given infra name
+func isExistingIdentifyProvider(client aws.Client, providerARN, infraName string) (bool, error) {
+	provider, err := client.GetOpenIDConnectProvider(&iam.GetOpenIDConnectProviderInput{
+		OpenIDConnectProviderArn: awssdk.String(providerARN),
+	})
+	if err != nil {
+		return false, errors.Wrapf(err, "Failed to get Identity Provider with ARN %s", providerARN)
+	}
+
+	for _, tag := range provider.Tags {
+		if tag.Key == awssdk.String(fmt.Sprintf("%s/%s", ccoctlAWSResourceTagKeyPrefix, infraName)) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func identityProviderCmd(cmd *cobra.Command, args []string) {
+	cfg := &awssdk.Config{
+		Region: awssdk.String(CreateOpts.Region),
+	}
+
+	s, err := session.NewSession(cfg)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	awsClient := &aws.AWSClient{
+		IAMClient: iam.New(s),
+		S3Client:  s3.New(s),
+	}
+
+	err = createIdentityProvider(awsClient, CreateOpts.InfraName, CreateOpts.Region, CreateOpts.PublicKeyPath, CreateOpts.TargetDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// NewIdentityProviderSetup provides the "create identity-provider" subcommand
+func NewIdentityProviderSetup() *cobra.Command {
+	identityProviderSetupCmd := &cobra.Command{
+		Use: "identity-provider",
+		Run: identityProviderCmd,
+	}
+
+	identityProviderSetupCmd.PersistentFlags().StringVar(&CreateOpts.InfraName, "infra-name", "", "Name prefix for all created AWS resources")
+	identityProviderSetupCmd.MarkPersistentFlagRequired("infra-name")
+	identityProviderSetupCmd.PersistentFlags().StringVar(&CreateOpts.Region, "region", "", "AWS region where the S3 OpenID Connect endpoint will be created")
+	identityProviderSetupCmd.MarkPersistentFlagRequired("region")
+	identityProviderSetupCmd.PersistentFlags().StringVar(&CreateOpts.PublicKeyPath, "public-key", "", "Path to public ServiceAccount signing key")
+	identityProviderSetupCmd.MarkPersistentFlagRequired("public-key")
```

Comment:
I believe we also want some kind of --dry-run or --skip-create flag so that the tool just places the AWS objects on the filesystem (so that a user can review and apply the objects if they don't trust ccoctl or they just want some extra tags, naming convention, etc).